### PR TITLE
Allow txindex in prune mode

### PR DESCRIPTION
**Pull request by jonasschnelli**
Idea is outlined here: #12651

This allows to enable the txindex when pruning is enabled.
During pruning a block file, the `CDiskTxPos` objects in the txindex do get migrated to a "per-blockheight" approach allowing to get the blockhash and tx-position of every txid.

Once #10794 (`requestblocks`, or similar) is merged, `getrawtransaction` could wait until the block is fetched via the connected peers and return the transaction. Retrieving a transaction would still be fast enough for non-high-load usage (less then a second to a couple of seconds).

This would allow to run a txindex with ~20GB of diskspace for non high-load usage like personal block explorers.

ToDo:
* [ ] Tests
* [ ] Release-Notes

### CODE REVIEW

File: src/validation.h [hunk 1]
```diff
@@ -278,7 +278,14 @@ void ThreadScriptCheck();
 /** Check whether we are doing an initial block download (synchronizing from disk or network) */
 bool IsInitialBlockDownload();
 /** Retrieve a transaction (from memory pool, or from disk, if possible) */
-bool GetTransaction(const uint256& hash, CTransactionRef& tx, const Consensus::Params& params, uint256& hashBlock, bool fAllowSlow = false, CBlockIndex* blockIndex = nullptr);
+enum class GetTransactionResult
+{
+    LOAD_OK,
```
Comment (Empact) [inline src/validation.h]: `OK` seems sufficient

**Comment by Sjors** [comment]
Does this work on top of #11857?

**Comment by jonasschnelli** [comment]
@sjors: I originally developed this on top of 11857... it's a trivial offset. So yes. It works on top of that.

File: src/validation.cpp [hunk 2]
```diff
@@ -1016,31 +1016,45 @@ bool GetTransaction(const uint256& hash, CTransactionRef& txOut, const Consensus
...
 
         if (fTxIndex) {
             CDiskTxPos postx;
             if (pblocktree->ReadTxIndex(hash, postx)) {
+                if (postx.nFile == std::numeric_limits<int>::max()) {
+                    // blockfile is not available (pruned), CDiskTxPos.nPos points to the blockheight
+                    // lookup blockindex via height from txindex and set hashBlock
+                    CBlockIndex *p_index_possible = chainActive[postx.nPos];
+                    if (p_index_possible) {
+                        hashBlock = *p_index_possible->phashBlock;
```
Comment (promag) [inline src/validation.cpp]: `hashBlock` is only updated if block index is found. Either `assert(hashBlock.IsNull())` before or just call `hashBlock.SetNull()`.

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -169,13 +169,23 @@ UniValue getrawtransaction(const JSONRPCRequest& request)
...
             if (!(blockindex->nStatus & BLOCK_HAVE_DATA)) {
                 throw JSONRPCError(RPC_MISC_ERROR, "Block not available");
             }
             errmsg = "No such transaction found in the provided block";
+        } else if (res == GetTransactionResult::BLOCK_PRUNED) {
+            //TODO: fetch block via peers, wait for download and return transaction
+            //see https://github.com/bitcoin/bitcoin/pull/10794
+            UniValue result(UniValue::VOBJ);
+            result.pushKV("status", "pruned");
+            result.pushKV("inblock", hash_block.GetHex());
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: `hash_block` can be null, should display it?

File: src/validation.cpp [hunk 4]
```diff
@@ -3544,13 +3558,41 @@ uint64_t CalculateCurrentUsage()
...
+                //        nTxOffset = index of transaction position in block (not to confused with data offset)
+                LogPrintf("txindex block pruning: %s migrate txindex data from block %s\n", __func__, pindex->GetBlockHash().GetHex());
+                CBlock block;
+                if (!ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+                    AbortNode(state, "Failed to read block");
+                    return;
+                }
+                std::vector<std::pair<uint256, CDiskTxPos>> vPos;
+                unsigned int txpos = 0;
+                for(const auto& tx : block.vtx)
```
Comment (promag) [inline src/validation.cpp]: Missing space after `for`. `{` in this line.

**Review by promag — COMMENTED** [review]
Concept ACK.

File: src/validation.cpp (hunk 4 above)
Comment (Empact) [reply src/validation.cpp]: clang-format will fix this and maybe more.

**Comment by TheBlueMatt** [comment]
Seems there's lots of debate on the utility here. Concept ACK/NACK discussion is at #12651, so should probably hold off review here so as to avoid splitting the discussion.

File: src/rpc/rawtransaction.cpp (hunk 3 above)
Comment (luke-jr) [inline src/rpc/rawtransaction.cpp]: Should be `"blockhash"`

**Comment by Sjors** [comment]
There should be tests to cover the scenario where:
1. an index exists and a block gets pruned
2. an already pruned node is launched with `-txindex` (this should probably fail)

The documentation should make it more clear that all transaction hashes since genesis are kept. Normally the index is about 17 GB on mainnet, does this PR make it smaller?

An alternative to this PR could be to drop old transactions from the index, though that reduces utility (perhaps better discussed in #12651).

Needs (not super difficult) rebase to work with the new `txindex` structure from #13033.

**Pull request closed without merging**